presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java

```diff
@@ -0,0 +1,132 @@
+import com.google.common.collect.ImmutableSet;
+
+import javax.inject.Inject;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import javax.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.Principal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.io.ByteStreams.copy;
+import static com.google.common.io.ByteStreams.nullOutputStream;
+import static com.google.common.net.HttpHeaders.WWW_AUTHENTICATE;
+import static javax.servlet.http.HttpServletResponse.SC_UNAUTHORIZED;
+
+public class AuthenticationFilter
+        implements Filter
+{
+    private final Set<Authenticator> authenticators;
+
+    @Inject
+    public AuthenticationFilter(Set<Authenticator> authenticators)
+    {
+        this.authenticators = ImmutableSet.copyOf(authenticators);
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) {}
+
+    @Override
+    public void destroy() {}
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain nextFilter)
+            throws IOException, ServletException
+    {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+        // skip authentication if non-secure or not configured
+        if (!request.isSecure() || authenticators.isEmpty()) {
+            nextFilter.doFilter(request, response);
+            return;
+        }
+
+        // try to authenticate, collecting errors and header values
+        List<String> messages = new ArrayList<>();
+        List<String> headerValues = new ArrayList<>();
+
+        for (Authenticator authenticator : authenticators) {
+            Principal principal;
+            try {
+                principal = authenticator.authenticate(request);
+            }
+            catch (AuthenticationException e) {
+                if (e.getMessage() != null) {
+                    messages.add(e.getMessage());
+                }
+                e.getHeaderValue().ifPresent(headerValues::add);
+                continue;
+            }
+
+            // authentication succeeded
+            nextFilter.doFilter(withPrincipal(request, principal), response);
+            return;
+        }
+
+        // authentication failed
+        skipRequestBody(request);
+
+        for (String value : headerValues) {
+            response.setHeader(WWW_AUTHENTICATE, value);
+        }
+
+        if (messages.isEmpty()) {
+            messages.add("Unauthorized");
+        }
+        response.sendError(SC_UNAUTHORIZED, Joiner.on(" | ").join(messages));
```

Comment:
This gets sent in the HTTP status which can't have a newline. That needs to change but is a different project. In practice, there should only ever be one message, since multiple messages would require the client to authenticate with multiple methods at once.